Track AV progress and send to google analytics

sarangj opened the pull request:
## Ticket:

- JIRA ticket [DR-3593](https://newyorkpubliclibrary.atlassian.net/browse/DR-3593)
- JIRA ticket [DR-3594](https://newyorkpubliclibrary.atlassian.net/browse/DR-3593)

## This PR does the following:

Use the plyr on 'timeupdate' event to track when specific milestones are reached when playing an AV file. The idea here is to track when a 'milestone' percentage gets crossed, and send an event for those milestones. I basically followed the same logic as in OG DC: https://github.com/NYPL/digitalreadingroom/blob/9c01d999ef8085b867869f65d87cbbc2bbaedeca/app/assets/javascripts/ga-plugin.js#L199-L217 where we wait until a given milestone is passed, but if we're more than 5% past the milestone, we don't trigger an event, to account for the case where a user is clicking around (ie, if the user skips straight to 40% into the video, we don't want to trigger an event for 25% progress reached).

Note, I had to split the plyr out into a child component to cleanly use the `usePlyr`, as in this example from the plyr-react library: https://github.com/chintan9/plyr-react/blob/eb579712a2e3a7b871a24daa7241acf01856699d/example/src/custom-audio-player.tsx#L18-L50

Also note that for the 100& milestone, I simply use the playback `end` event.

## Open questions

<!-- Any questions you want to ask the reviewer? -->

## How has this been tested? How should a reviewer test this?

I basically tested this by playing some AV material locally and inspecting console logs, this will need to be further tested in QA.

## Accessibility concerns or updates

<!--- Describe any accessibility concerns or updates that were made that should be known. -->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.
- [ ] I have updated the CHANGELOG.md.


[DR-3593]: https://newyorkpubliclibrary.atlassian.net/browse/DR-3593?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DR-3594]: https://newyorkpubliclibrary.atlassian.net/browse/DR-3594?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sarangj committed:
Track AV progress and send to google analytics

Use the plyr on 'timeupdate' event to track when specific milestones are
reached when playing an AV file. The idea here is to track when a
'milestone' percentage gets crossed, and send an event for those
milestones. I basically followed the same logic as in OG DC:
https://github.com/NYPL/digitalreadingroom/blob/9c01d999ef8085b867869f65d87cbbc2bbaedeca/app/assets/javascripts/ga-plugin.js#L199-L217
where we wait until a given milestone is passed, but if we're more than
5% past the milestone, we don't trigger an event, to account for the
case where a user is clicking around (ie, if the user skips straight to
40% into the video, we don't want to trigger an event for 25% progress
reached).

Note, I had to split the plyr out into a child component to cleanly use
the `usePlyr`, as in this example from the plyr-react library:
https://github.com/chintan9/plyr-react/blob/eb579712a2e3a7b871a24daa7241acf01856699d/example/src/custom-audio-player.tsx#L18-L50

Also note that for the 100& milestone, I simply use the playback `end`
event.

sarangj commented on app/src/components/items/plyr/player.tsx:
```diff
@@ -74,20 +75,10 @@ const Player = ({ title, sources, type }: PlyrProps) => {
-          height={playerHeight}
-          width="100%"
-        />
+        <CustomPlyr ref={plyrRef} source={source} options={undefined} />
```
I don't need to pass `options` here

sarangj commented on app/src/components/items/plyr/player.tsx:
```diff
@@ -115,4 +106,43 @@ const Player = ({ title, sources, type }: PlyrProps) => {
+    };
+    const api = current as { plyr: PlyrInstance };
+    api.plyr.on("timeupdate", handleTimeUpdate);
+    api.plyr.on("ended", () => trackAVProgress(source.type, source.title, 100));
```
I should make sure this only fires once in case the user navigates back and ends the playback multiple times.

sarangj commented on app/src/components/items/plyr/player.tsx:
```diff
@@ -115,4 +106,43 @@ const Player = ({ title, sources, type }: PlyrProps) => {
+    api.plyr.on("ended", () => trackAVProgress(source.type, source.title, 100));
+  });
+  return (
+    <video
```
This seems to work for audio, but I suppose I should do this correctly

sarangj replied:
hmmm actually this seems to be the way - the ply-react provided audio example still uses a video element: https://github.com/chintan9/plyr-react/blob/eb579712a2e3a7b871a24daa7241acf01856699d/example/src/custom-audio-player.tsx#L43-L46

sarangj committed:
Do not double log media end

sarangj commented:
i've been testing this audio for anyone who wants to look: https://digital-collections-git-videoanalytics-nypl.vercel.app/items/aa56ae00-b884-0133-3277-60f81dd2b63c

you'll see the milestone percentages logged in the console

avertrees commented on app/src/components/items/plyr/player.tsx:
```diff
@@ -115,4 +106,48 @@ const Player = ({ title, sources, type }: PlyrProps) => {
   );
 };
 
+const CustomPlyr = forwardRef<APITypes, any>((props, ref) => {
```
nit but can we make this it's own separate component?

sarangj replied:
done!

sarangj committed:
Merge remote-tracking branch 'origin/qa' into video_analytics

sarangj committed:
Update AV playback logging

Add a log statement for when playback ends and add some context around
the percentage milestones

sarangj committed:
Extract new component into its own module

To keep things clean, open to feedback on naming
